src/bundle.c

```diff
@@ -132,66 +103,129 @@ bool is_installed_bundle(const char *bundle_name)
 	return ret;
 }
 
-/* When loading all tracked bundles into memory, they happen
- * to be hold in subs global var, for some reasons it is
- * needed to just pop out one or more of this loaded tracked
- * bundles, this function search for bundle_name into subs
- * struct and if it found then free it from the list.
- */
-static enum swupd_code unload_tracked_bundle(const char *bundle_name, struct list **subs)
+static int find_manifest(const void *a, const void *b)
 {
-	struct list *bundles;
-	struct list *cur_item;
-	struct sub *bundle;
-
-	bundles = list_head(*subs);
-	while (bundles) {
-		bundle = bundles->data;
-		cur_item = bundles;
-		bundles = bundles->next;
-		if (strcmp(bundle->component, bundle_name) == 0) {
-			/* unlink (aka untrack) matching bundle name from tracked ones */
-			*subs = free_bundle(cur_item);
-			return SWUPD_OK;
-		}
+	struct manifest *A;
+	char *B;
+	int ret;
+
+	A = (struct manifest *)a;
+	B = (char *)b;
+
+	ret = strcmp(A->component, B);
+	if (ret != 0) {
+		return ret;
 	}
 
-	return SWUPD_BUNDLE_NOT_TRACKED;
+	/* we found a match*/
+	return 0;
 }
 
 /* Return list of bundles that include bundle_name */
-static void required_by(struct list **reqd_by, const char *bundle_name, struct manifest *mom, int recursion)
+static int required_by(struct list **reqd_by, const char *bundle_name, struct manifest *mom, int recursion, struct list *exclusions, char *msg)
 {
 	struct list *b, *i;
+	char *name;
+	int count = 0;
+	static bool print_msg;
+	bool verbose = (log_get_level() == LOG_INFO_VERBOSE);
+
 	// track recursion level for indentation
+	if (recursion == 0) {
+		print_msg = true;
+	}
 	recursion++;
 
+	/* look at the manifest of all listed bundles to see if
+	 * they list *bundle_name as a dependency */
 	b = list_head(mom->submanifests);
 	while (b) {
 		struct manifest *bundle = b->data;
 		b = b->next;
 
+		if (strcmp(bundle->component, bundle_name) == 0) {
+			/* circular dependencies are not allowed in manifests,
+			 * so we can skip checking for dependencies in the same bundle */
+			continue;
+		}
+
 		int indent = 0;
 		i = list_head(bundle->includes);
 		while (i) {
-			char *name = i->data;
+			name = i->data;
 			i = i->next;
 
 			if (strcmp(name, bundle_name) == 0) {
-				char *bundle_str = NULL;
-				indent = (recursion - 1) * 4;
-				if (recursion == 1) {
-					string_or_die(&bundle_str, "%*s* %s\n", indent + 2, "", bundle->component);
-				} else {
-					string_or_die(&bundle_str, "%*s|-- %s\n", indent, "", bundle->component);
+
+				/* this bundle has *bundle_name as a dependency */
+
+				/* if the bundle being looked at is in the list of exclusions
+				 * then don't consider it as a dependency, the user added it to
+				 * the list of bundles to be removed too, but we DO want to
+				 * consider its list of includes */
+				if (list_search(exclusions, bundle->component, list_strcmp)) {
+					goto check_includes;
```

Comment:
Avoid gotos unless you're checking for errors or going to the end of a function.

Reply:
ok, I'll replace this goto with an if 👌